subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/serialization/Combinators.kt

```diff
@@ -279,27 +279,51 @@ internal
 fun Encoder.writeClassPath(classPath: ClassPath) {
     // Ensure that the proper type is going to be restored,
     // because it is important for the equality checks.
-    writeBoolean(classPath is TransformedClassPath)
-    writeCollection(classPath.asFiles) {
-        writeFile(it)
+    if (classPath is TransformedClassPath) {
+        writeBoolean(true)
+        writeCollection(classPath.asFiles.zip(classPath.asTransformedFiles)) {
```

Comment:
🎨 At this point I'd extract a `writeTransformedClassPath` and a `writeDefaultClassPath` for symmetry.